### PR TITLE
Scrapper, Anastasia Logunova - 22FPL1

### PR DESCRIPTION
**Pull request by logunova-anastasia**

### CODE REVIEW

**Comment by demid5111** [comment]
nothing to review

File: requirements.txt [hunk 1]
```diff
@@ -1 +1,5 @@
-
+beautifulsoup4==4.12.0
+mypy==1.9.0
```
Comment (demid5111) [inline requirements.txt:2]: requirements should include only those libraries that fuel work of your lab. test and check related things should not be present here

File: lab_5_scrapper/scrapper_config.json [hunk 2]
```diff
@@ -1,9 +1,11 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": ["https://scientificrussia.ru/news", "https://scientificrussia.ru/news/1224", "https://scientificrussia.ru/news/1223"],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json:2]: are these urls enough to collect 100 articles?

File: lab_5_scrapper/scrapper_config.json [hunk 3]
```diff
@@ -1,9 +1,11 @@
...
+    "headers": {"Accept": "*/*",
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json:3]: apply better formatting to JSON. for example, you can see how it should be formatted: https://jsonformatter.curiousconcept.com/#

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -204,12 +344,26 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    if not base_path.exists():
+        base_path.mkdir()
+    base_path.rmdir()
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:349]: if it is a directory with content, then it will not be so easy

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -176,17 +303,23 @@ def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> No
...
-            datetime.datetime: Datetime object
-        """
+        author = [' '.join(article_soup.find(class_='author').text.split()[1:])]
+        if author:
+            self.article.author = author
+        else:
+            self.article.author = ["NOT FOUND"]
+        self.article.title = article_soup.find(itemprop="name headline").text.strip()
+
+    # def unify_date_format(self, date_str: str) -> datetime.datetime:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:313]: uncomment

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -168,6 +290,11 @@ def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
         Args:
             article_soup (bs4.BeautifulSoup): BeautifulSoup instance
         """
+        text = article_soup.find_all(style="text-align: justify;")
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:293]: looks like too broad condition - you will get some more data than you need

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -168,6 +290,11 @@ def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
...
+        article = ''
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:294]: use more efficient concatenation strategy

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -160,6 +278,10 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
             article_id (int): Article id
             config (Config): Configuration
         """
+        self.full_url = full_url
+        self.article_id = article_id
+        self.config = config
+        self.article = Article(self.full_url, self.article_id)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:284]: is it stated in docs to be an attribute?

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -127,11 +227,29 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+                url = link['href']
+        return self.url_pattern + url
 
     def find_articles(self) -> None:
         """
         Find articles.
         """
+        urls = []
+
+        for url in self.get_search_urls():
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:242]: how do you control number of articles to collect? they are defined in config

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -127,11 +227,29 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+            response = make_request(url, self.config)
+
+            if not response.ok:
+                continue
+
+            src = response.text
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:248]: no need in this variable

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -127,11 +227,29 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
         Returns:
             str: Url from HTML
         """
+        url = ''
+        for div in article_bs.find_all("div", {"class": "title"}, limit=24):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:231]: why do you have this limit?

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -26,11 +77,47 @@ def _extract_config_content(self) -> ConfigDTO:
...
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters are not corrupt.
         """
+        if not isinstance(self.config_dto.seed_urls, list):
+            raise IncorrectSeedURLError
+
+        for seed_url in self.config_dto.seed_urls:
+            if not re.match(r"https?://(www.)?", seed_url):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:100]: we do not work with any website, right? we need seed urls to be exactly of the website we target

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -26,11 +77,47 @@ def _extract_config_content(self) -> ConfigDTO:
         Returns:
             ConfigDTO: Config values
         """
+        with open(self.path_to_config, 'r', encoding='utf-8') as file:
+            config = json.load(file)
+
+        return ConfigDTO(
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:83]: imagine how to do it with kwargs-style

**Review by demid5111 — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 1 above)
Comment (logunova-anastasia) [reply requirements.txt:2]: removed it

File: lab_5_scrapper/scrapper_config.json (hunk 2 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper_config.json:2]: nope, forgot to change it. done!

File: lab_5_scrapper/scrapper.py (hunk 7 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:294]: changed it

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:242]: done

File: lab_5_scrapper/scrapper.py (hunk 10 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:248]: done

File: lab_5_scrapper/scrapper.py (hunk 13 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:83]: did it

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:349]: changed it

File: lab_5_scrapper/scrapper_config.json (hunk 3 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper_config.json:3]: did it!

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:349]: fixed it

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:313]: done

File: lab_5_scrapper/scrapper.py (hunk 11 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:231]: it was a maximum number of articles per page, but I figured out how to do it without this limit

File: lab_5_scrapper/scrapper.py (hunk 12 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:100]: fixed it

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:293]: i tried it and got a text only from an article

Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:293]: the only thing is that i get some information about a source of a photo, but it is present in articles and located in articleBody, should i get rid of it?

File: lab_5_scrapper/scrapper.py [hunk 14]
```diff
@@ -2,9 +2,51 @@
...
+from bs4 import BeautifulSoup
+
+from core_utils.article.article import Article
+from core_utils.article.io import to_meta, to_raw
+from core_utils.config_dto import ConfigDTO
+from core_utils.constants import ASSETS_PATH, CRAWLER_CONFIG_PATH
+
+
+class IncorrectSeedURLError(Exception):
+    pass
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:24]: provide docstrings to custom error classes

File: lab_5_scrapper/scrapper.py [hunk 15]
```diff
@@ -127,11 +222,27 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
         Returns:
             str: Url from HTML
         """
+        url = ''
+        for div in article_bs.find(class_="card-body").find_all("div", {"class": "title"}):
+            for link in div.select("a"):
+                url = link['href']
+        return self.url_pattern + url
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:229]: what is the reason to loop over ALL urls and take only LAST?

File: lab_5_scrapper/scrapper.py [hunk 16]
```diff
@@ -127,11 +222,27 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+        urls = []
+        while len(urls) < self.config.get_num_articles():
+            for url in self.get_search_urls():
+                response = make_request(url, self.config)
+
+                if not response.ok:
+                    continue
+
+                soup = BeautifulSoup(response.text, 'lxml')
+                urls.append(self._extract_url(soup))
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:244]: how do you check for duplications?

File: lab_5_scrapper/scrapper.py [hunk 17]
```diff
@@ -127,11 +222,27 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+        self.urls.extend(urls)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:245]: why not changing `self.urls` directly? What is the reason of first using intermediate variable to store links?

File: lab_5_scrapper/scrapper.py [hunk 18]
```diff
@@ -195,6 +338,12 @@ def parse(self) -> Union[Article, bool, list]:
         Returns:
             Union[Article, bool, list]: Article instance
         """
+        response = make_request(self.full_url, self.config)
+        article_bs = BeautifulSoup(response.text, 'lxml')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:342]: who guaranteed that the request is always success?

File: requirements.txt [hunk 19]
```diff
@@ -1 +1,4 @@
-
+beautifulsoup4==4.12.0
+lxml==4.9.2
+requests==2.31.0
+stanza==1.8.1
```
Comment (demid5111) [inline requirements.txt:4]: rm

File: lab_5_scrapper/scrapper.py [hunk 20]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+        parser = HTMLParser(url, id_num, conf)
+        article = parser.parse()
+        if isinstance(article, Article):
+            to_raw(article)
+            to_meta(article)
 
 
 if __name__ == "__main__":
-    main()
+    recursive_main()
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:466]: keep main. it is better for CI. for demonstration in class, you will change what you call.

File: lab_5_scrapper/scrapper.py [hunk 21]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+            self.visited_urls = data['visited_urls']
+
+    def save_info(self) -> None:
+        data = {
+            'num_visited_urls': self.num_of_urls,
+            'urls': self.urls,
+            'possible_urls': self.possible_urls,
+            'visited_urls': self.visited_urls
+        }
+        with open(pathlib.Path(__file__).parent.parent / 'tmp' / 'recursive_crawler.json', 'w',
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: this is the thrid time I see this path. how about DRY?

**Review by demid5111 — CHANGES_REQUESTED** [review]
super! you have green CI and solid draft of recursive crawler. one of best works in my group to this moment. so, keep going!

File: lab_5_scrapper/scrapper.py (hunk 15 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py]: fixed it

File: lab_5_scrapper/scrapper.py (hunk 16 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py]: fixed

File: lab_5_scrapper/scrapper.py (hunk 17 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py]: changed it

File: requirements.txt (hunk 19 above)
Comment (logunova-anastasia) [reply requirements.txt:4]: removed

File: lab_5_scrapper/scrapper.py (hunk 20 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py]: did it

File: lab_5_scrapper/scrapper.py [hunk 22]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+        super().__init__(config)
+        self.start_url = self.config.get_seed_urls()[0]
+        self.urls = []
+        self.num_of_urls = 1
+        self.possible_urls = [self.start_url]
+        self.visited_urls = []
+        self.get_info()
+
+    def get_info(self) -> None:
+        if (pathlib.Path(__file__).parent.parent / 'tmp' / 'recursive_crawler.json').exists():
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py]: did it

File: lab_5_scrapper/scrapper.py (hunk 21 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py]: changed it

File: lab_5_scrapper/scrapper.py [hunk 23]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+
+            with open(pathlib.Path(__file__).parent.parent / 'tmp' / 'recursive_crawler.json', 'r', encoding='utf-8') as infile:
+                data = json.load(infile)
+
+            self.urls = data['urls']
+            self.num_of_urls = data['num_visited_urls']
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:385]: yes, changed it

File: lab_5_scrapper/scrapper.py [hunk 24]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+        for div in article_bs.find(class_="card-body").find_all("div", {"class": "title"}):
+            for link in div.select("a"):
+                articles.append(link['href'])
+
+        for i in links:
+            if i.get('href'):
+                if 'https' in i.get('href'):
+                    link_url = i.get('href')
+                else:
+                    link_url = 'https://scientificrussia.ru' + i.get('href')
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:420]: previously in sequential crawler i was searching for links with article titles only and there was no need for this(as i was getting only links looking like this:  href="/articles/...", but here i search for all links and therefore get results like 'href="https://vk.com/scientificrussia"'

File: lab_5_scrapper/scrapper.py (hunk 18 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:344]: fixed

File: lab_5_scrapper/scrapper.py [hunk 25]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+            json.dump(data, file, indent=4)
+
+    def find_articles(self) -> None:
+        url = self.possible_urls[self.num_of_urls-1]
+
+        if url not in self.visited_urls:
+            self.visited_urls.append(url)
+
+        response = make_request(url, self.config)
+        article_bs = BeautifulSoup(response.text, 'lxml')
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:407]: fixed

File: lab_5_scrapper/scrapper.py (hunk 14 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:24]: did it

File: lab_5_scrapper/scrapper.py [hunk 26]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+        """
+        super().__init__(config)
+        self.start_url = self.config.get_seed_urls()[0]
+        self.urls = []
+        self.num_of_urls = 1
+        self.possible_urls = [self.start_url]
+        self.visited_urls = []
+        self.get_info()
+
+    def get_info(self) -> None:
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:379]: done

File: lab_5_scrapper/scrapper.py [hunk 27]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    if base_path.exists():
+        shutil.rmtree(base_path)
+    base_path.mkdir(parents=True)
+
+
+class CrawlerRecursive(Crawler):
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:363]: done

File: lab_5_scrapper/scrapper.py [hunk 28]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+    def __init__(self, config: Config) -> None:
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:364]: done

File: lab_5_scrapper/scrapper.py [hunk 29]
```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+            self.visited_urls.append(url)
+
+        response = make_request(url, self.config)
+        article_bs = BeautifulSoup(response.text, 'lxml')
+
+        links = article_bs.find_all('a')
+        articles = []
+        for div in article_bs.find(class_="card-body").find_all("div", {"class": "title"}):
+            for link in div.select("a"):
+                articles.append(link['href'])
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py]: applied extract_url here

File: lab_5_scrapper/scrapper.py [hunk 30]
```diff
@@ -204,12 +376,126 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+
+        for id_num, url in enumerate(crawler.urls, 1):
+            parser = HTMLParser(url, id_num, conf)
+            article = parser.parse()
+            if isinstance(article, Article):
+                to_raw(article)
+                to_meta(article)
+
+    except KeyboardInterrupt:
+        crawler.save_info()
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:498]: strange save strategy. imagine machine is dead. you won't save anything in current implementation - not all articles are collected and it is not keyboard interrupt

**Review by demid5111 — CHANGES_REQUESTED** [review]
nice, some polishing is required.

File: lab_5_scrapper/scrapper.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:384]: done

File: lab_5_scrapper/scrapper.py [hunk 31]
```diff
@@ -127,11 +243,27 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
         Returns:
             str: Url from HTML
         """
+        if article_bs.find('a').get('href'):
+            return self.url_pattern + str(article_bs.find('a').get('href'))
+        return ''
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:248]: done

File: lab_5_scrapper/scrapper.py [hunk 32]
```diff
@@ -204,12 +376,126 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+        self.possible_urls = [self.start_url]
+        self.path = ASSETS_PATH.parent / 'recursive_crawler.json'
+        self.visited_urls = []
+        self.get_info()
+
+    def get_info(self) -> None:
+        """
+        Get information about the crawler from a file.
+        """
+        if self.path.exists():
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:404]: done

File: lab_5_scrapper/scrapper.py [hunk 33]
```diff
@@ -204,12 +376,126 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+        article_bs = BeautifulSoup(response.text, 'lxml')
+
+        links = article_bs.find_all('a')
+        articles = []
+
+        for div in article_bs.find(class_="card-body").find_all("div", {"class": "title"}):
+            articles.append(self._extract_url(div))
+
+        for i in links:
+            if i.get('href'):
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:448]: done

File: lab_5_scrapper/scrapper.py (hunk 30 above)
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:498]: yes, didn't consider it, changed the strategy

File: lab_5_scrapper/scrapper.py [hunk 34]
```diff
@@ -204,12 +376,126 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
...
+            self.possible_urls = data['possible_urls']
+            self.visited_urls = data['visited_urls']
+
+    def save_info(self) -> None:
+        """
+        Save information about the crawler.
+        """
+        data = {
+            'urls': self.urls,
+            'number of collected urls': len(self.urls),
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_5_scrapper/scrapper.py:418]: just want it to be visible

**Pull request closed without merging**